feat: [DHIS213343] hidden program stage rule effect

simonadomnisoru opened the pull request:
[DHIS2-13343](https://dhis2.atlassian.net/browse/DHIS2-13343)

**Tech summary**

`/enrollment` page

- Quick actions → new and schedule event: Rule effect might disable the buttons 
- Stages and events → new [stagename] event: Rule effect might disable the button

`/enrollmentEventNew` page
- Stage not selected: Rule effect might disable stage buttons. Start running rules in the program stage selector.
- Stage selected: Show error if hidden program stage rule is in effect and disable the form action buttons

superskip commented on cypress/integration/EnrollmentPage/HiddenProgramStage/index.js:
```diff
@@ -0,0 +1,65 @@
+
+    cy.contains('[data-test="program-stage-selector-button"]', 'Postpartum care visit').should('be.disabled');
+
+    // clean up
+    cy.visit(
+        '/#/enrollment?enrollmentId=fmhIsWXVDmS&orgUnitId=s7SLtx8wmRA&programId=WSGAb5XwJ3Y&teiId=uW8Y7AIcRKA',
+    );
+
+    cy.get('[data-test="enrollment-page-content"]').contains('Enrollment Dashboard');
+
+    cy.get('[data-test="stages-and-events-widget"]')
+        .find('[data-test="stage-content"]')
+        .eq(3)
+        .click();
+
+    cy.contains('WHOMCH Pregnancy outcome').should('exist');
+    cy.contains('[data-test="dhis2-uicore-button"]', 'Edit event').click();
+    cy.contains('[data-test="dhis2-uicore-button"]', 'Delete').click();
+    cy.contains('[data-test="dhis2-uicore-button"]', 'Yes, delete event').click();
```
Relying on the test to do cleanup at the end might cause problems if the test fails mid-way. If cleanup is required then I've understood that best practice is to do it at the beginning of a test.

superskip reviewed (commented):
Seems like you've made a good job figuring out edge cases! :+1: 

I noticed that the design doc mentions here and there to add a tooltip message “You can’t add any more $ProgramStageName events", doesn't look like this has been implemented yet. I also added a comment on the cypress test. :innocent:

simonadomnisoru commented:
> Seems like you've made a good job figuring out edge cases! 👍
> 
> I noticed that the design doc mentions here and there to add a tooltip message “You can’t add any more $ProgramStageName events", doesn't look like this has been implemented yet. I also added a comment on the cypress test. 😇

Hey @superskip,
Good catch! I added the missing tooltips.
Thank you for the feedback!

superskip reviewed (commented):
Tooltip implementation looks good :+1:
Just one more concern with the test:

simonadomnisoru commented on cypress/integration/EnrollmentPage/HiddenProgramStage/index.js:
```diff
@@ -1,7 +1,25 @@
+    cy.get('[data-test="stages-and-events-widget"]')
+        .find('[data-test="stage-content"]')
+        .eq(3)
+        .click();
```
Great suggestion! I updated to code to handle the case of no events in the program stage. Thanks

superskip reviewed (approved):
All good! :star:

geethaalwan reviewed (approved):
Tested successfully on 2.41,2.40.2,2.39.3,2.38.5 versions
